octobot_commons/databases/cache_manager.py

```diff
@@ -0,0 +1,223 @@
+                                       tentacle_name, identifier, tentacles_setup_config,
+                                       cache_type, tentacles_requirements):
+        cache_full_path = self.get_cache_or_build_path(
+            tentacle, exchange, symbol, time_frame,
+            tentacle_name, identifier, tentacles_setup_config, tentacles_requirements)
+        cache_dir = os.path.split(cache_full_path)[0]
+        if not os.path.exists(cache_dir):
+            os.makedirs(cache_dir)
+        return self._open_cache_database(cache_full_path, cache_type, tentacles_requirements)
+
+    def _open_cache_database(self, file_path, cache_type, tentacles_requirements):
+        """
+        Override to use another cache database or adaptor
+        :return: the cache database class
+        """
+        return _CacheWrapper(file_path, cache_type, self.database_adaptor, tentacles_requirements)
+
+    def get_cache_or_build_path(self, tentacle, exchange, symbol, time_frame, tentacle_name, config_name,
+                                tentacles_setup_config, tentacles_requirements):
+        identifier = config_name or self.DEFAULT_CONFIG_IDENTIFIER
+        try:
+            return self.__class__.CACHES[tentacle_name][exchange][symbol][time_frame][identifier].get_path()
+        except KeyError:
+            sanitized_pair = symbol_util.merge_symbol(symbol) if symbol else symbol
+            required_tentacles = tentacles_requirements.get_all_required_tentacles(False)
+            # ensure tentacles requirements are snapshotting the configuration that was used to build the
+            # cache identifier
+            tentacles_requirements.synchronize_tentacles_config()
+            identifying_tentacles = [tentacle] + required_tentacles
+            # warning: very slow, should be called as rarely as possible
+            code_hash, config_hash = self._tentacles_hashes(identifying_tentacles, tentacles_setup_config)
+            return os.path.join(common_constants.USER_FOLDER, common_constants.CACHE_FOLDER, tentacle_name,
+                                exchange, sanitized_pair, time_frame,
+                                code_hash, config_hash,
+                                common_constants.CACHE_FILE)
+
+    @staticmethod
+    def _tentacles_hashes(identifying_tentacles, tentacles_setup_config) -> (str, str):
+        try:
+            import octobot_tentacles_manager.api
+            return \
+                octobot_tentacles_manager.api.get_code_hash(identifying_tentacles)[:common_constants.CACHE_HASH_SIZE], \
+                octobot_tentacles_manager.api.get_config_hash(
+                    identifying_tentacles,
+                    tentacles_setup_config
+                )[:common_constants.CACHE_HASH_SIZE]
+        except ImportError as e:
+            raise ImportError("octobot_tentacles_manager is required to use cache") from e
+
+
+class _CacheWrapper:
```

Comment:
Is it mandatory to have this class in the same file as CacheManager?

Reply:
Nope, I'll extract it 👍